### PR TITLE
Psa additional tests

### PR DESCRIPTION
**Pull request by anupama2501**
## Issue: 
https://github.com/rancher/qa-tasks/issues/680
 
## Problem
This PR addresses the additional tests required to validate PSA test cases. 

## Solution
We have a new custom role that needs to be validated. Created a new role updatePSA for this validation and this role should allow users to update, create and remove PSA labels in the test. 
 

### Automated Testing
Jenkins test run.

This PR depends slightly on the commit  https://github.com/rancher/rancher/pull/41560/commits/23250d2a0582dc06493a354ca2aa4c0a9a192a29, so the PR needs to be rebased once the changes for the linked commit were merged.

### CODE REVIEW

File: tests/v2/validation/rbac/rbac_psa_test.go [hunk 1]
```diff
@@ -291,12 +314,35 @@ func (rb *PSATestSuite) TestPSA() {
 		})
 
 		rb.Run("Testcase - Validate if members with roles "+role+"can add/edit/delete labesl from admin created namespace", func() {
-			rb.ValidatePSA(role)
+			if role == roleCustomCreateNS {
+				customRole = true
+			} else {
```
Comment (brudnak) [inline tests/v2/validation/rbac/rbac_psa_test.go:319]: Tiny nit: Optionally, consider removing else for cleaner code. By defaulting customRole to false and changing only if needed, the code could be slightly more readable.

```go
customRole = false
if role == roleCustomCreateNS {
    customRole = true
}
```

File: tests/v2/validation/rbac/rbac_psa_test.go [hunk 2]
```diff
@@ -291,12 +314,35 @@ func (rb *PSATestSuite) TestPSA() {
...
+			})
+		}
+
+		if strings.Contains(role, "project") || role == roleCustomCreateNS {
+			rb.Run("Additional testcase - Validate if "+role+" with an additional role update-psa can add/edit/delete labels from admin created namespace", func() {
+				err := users.AddClusterRoleToUser(rb.client, rb.cluster, rb.nonAdminUser, rb.psaRole.ID)
+				require.NoError(rb.T(), err)
+				if role == roleCustomCreateNS {
+					customRole = true
+				} else {
```
Comment (brudnak) [inline tests/v2/validation/rbac/rbac_psa_test.go:342]: Same as previously commented:

Tiny nit: Optionally, consider removing else for cleaner code. By defaulting customRole to false and changing only if needed, the code could be slightly more readable.

```go
customRole = false
if role == roleCustomCreateNS {
    customRole = true
}
```

File: tests/v2/validation/rbac/rbac_test.go [hunk 3]
```diff
@@ -67,14 +67,19 @@ func (rb *RBTestSuite) ValidateListCluster(role string) {
...
 	err = v1.ConvertToK8sType(clusterList.Data[0].Status, clusterStatus)
 	require.NoError(rb.T(), err)
 
-	actualClusterID := clusterStatus.ClusterName
-	assert.Equal(rb.T(), rb.cluster.ID, actualClusterID)
+	if role != restrictedAdmin {
+		assert.Equal(rb.T(), 1, len(clusterList.Data))
+		actualClusterID := clusterStatus.ClusterName
+		assert.Equal(rb.T(), rb.cluster.ID, actualClusterID)
+	} else {
```
Comment (brudnak) [inline tests/v2/validation/rbac/rbac_test.go:78]: Tiny optional style nit: Consider flipping the if condition and returning early to reduce nesting and increase code readability. Handling the restrictedAdmin role at the start and then returning makes it clear what special actions are taken for this role. Would that work?

```go
	if role == restrictedAdmin {
		adminClusterList, err := rb.client.Steve.SteveType(clusters.ProvisioningSteveResouceType).ListAll(nil)
		require.NoError(rb.T(), err)
		assert.Equal(rb.T(), (len(adminClusterList.Data) - 1), len(clusterList.Data))
		return
	}

	assert.Equal(rb.T(), 1, len(clusterList.Data))
	actualClusterID := clusterStatus.ClusterName
	assert.Equal(rb.T(), rb.cluster.ID, actualClusterID)
}
```

**Review by brudnak — APPROVED** [review]
lgtm - added some small optional style nits in the comments

**Review by caliskanugur — COMMENTED** [review]
Same as Andrew's nits, looks good to me

File: tests/v2/validation/rbac/rbac_psa_test.go (hunk 1 above)
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_psa_test.go:319]: agreed

File: tests/v2/validation/rbac/rbac_psa_test.go (hunk 2 above)
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_psa_test.go:342]: agreed

File: tests/v2/validation/rbac/rbac_test.go (hunk 3 above)
Comment (igomez06) [reply tests/v2/validation/rbac/rbac_test.go:78]: agreed

**Review by igomez06 — COMMENTED** [review]
looks good, had the same nits as Andrew.

**Review by slickwarren — COMMENTED** [review]
lgtm, same comments as above

File: tests/v2/validation/rbac/rbac_psa_test.go (hunk 1 above)
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac_psa_test.go]: thanks Andrew! nice suggestion. yep the keyword is initialized at L257.

File: tests/v2/validation/rbac/rbac.go [hunk 4]
```diff
@@ -160,9 +162,9 @@ func createDeploymentAndWait(steveclient *v1.Client, client *rancher.Client, clu
 		if *deployment.Spec.Replicas == deployment.Status.AvailableReplicas {
 			return true, nil
 		}
-		return false, nil
+		return false, err
```
Comment (igomez06) [inline tests/v2/validation/rbac/rbac.go:165]: Are you sure you want to return an error here? If there's an error the poll will stop and be reported as a failure.

**Review by igomez06 — COMMENTED** [review]
Looks great, one concern about the poll

File: tests/v2/validation/rbac/rbac.go (hunk 4 above)
Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go:165]: yeah so, the reason why I added this err was in the L156, when I deploy something with restricted PSAs enabled on the namespace, the deployment should fail and the error message needs to be returned. If I make the err nil in the return value, the forbidden error that I get from the failed deployment is not returned and the validations in the tests/v2/validation/rbac/rbac_psa_test.go on line L254 fail as an error is expected there.

Comment (igomez06) [reply tests/v2/validation/rbac/rbac.go:165]: Will this affect our other rbac tests? What if the availableReplicas don't match the replicas on the first loop?

Comment (anupama2501) [reply tests/v2/validation/rbac/rbac.go:165]: oh wait. sorry my bad! I thought this was outside of the polling function. Actually this is not necessary. The one in L167 is the one I wanted. this was a typo on my part. thanks for pointing that out.

**Review by slickwarren — APPROVED** [review]
lgtm

**Comment by slickwarren** [comment]
can you also rebase?

**Comment by igomez06** [comment]
> can you also rebase?

yes please rebase, and we'll get this merged in once drone checks pass

**Comment by anupama2501** [comment]
Rebased the PR and the drone run has passed. can we merge this?

**Review by brudnak — APPROVED** [review]

**Pull request merged**